### PR TITLE
[SE-0365] Fix #64757: Unexpected "Implicit use of 'self' in closure" error in closure nested in result builder

### LINKED ISSUE #64757: closure capture rule applied at Swift 5.8 doesn't work correctly in resultBuilder
**Description**
I acknowledged capturing rule became more strict with [SE-0365](https://github.com/apple/swift-evolution/blob/main/proposals/0365-implicit-self-weak-capture.md). When I tried to compile with Xcode 14.3 RC 2, our company's project codes didn't compile with some reasons. One of that is this case. In resultBuilder closure, using weak capture self with [if-let shorthand](https://github.com/apple/swift-evolution/blob/main/proposals/0345-if-let-shorthand.md) make capturing semantic error. I made a sample code snippet to reproduce problem.

**Steps to reproduce**
Try this sample

```swift
@resultBuilder
struct VoidBuilder {
    static func buildBlock() -> Void {

    }

    static func buildPartialBlock<T>(first: T) -> Void {

    }

    static func buildPartialBlock<T>(accumulated: Void, next: T) -> Void {

    }
}

final class EscapingWrapper {
    static func wrapper(_ closure: @escaping () -> Void) {
        closure()
    }
}

final class Sample {

    @VoidBuilder
    var void: Void {
        EscapingWrapper.wrapper { [weak self] in
            guard let self else { return } // Implicit use of 'self' in closure; use 'self.' to make capture semantics explicit
        }
    }
}
```

**Expected behavior**
compile success without error. 

**Environment**
- Swift compiler version info
swift-driver version: 1.75.2 Apple Swift version 5.8 (swiftlang-5.8.0.124.2 clang-1403.0.22.11.100)
Target: arm64-apple-macosx13.0

- Xcode version info
Xcode 14.3
Build version 14E222b
- Deployment target
 no specific target(It reproduced in both ios project and cmd project)

### LINKED ISSUE #71040: Compiler crash with if else expression in result builder (composable architecture reducer body)
### Description

Using if else expressions of Swift 5.9 in a composable architecture reducer always fails to compile for me as the compiler crashes. The following minimal sample code reproduces this for the latest Xcode in a new and clean project.

[TCAIfElseExpression.zip](https://github.com/apple/swift/files/13999772/TCAIfElseExpression.zip)

### Reproduction

```swift
import ComposableArchitecture

@Reducer
struct Feature {
  struct State {
    var value: Bool
  }
  enum Action {
    case dummy
  }
  var body: some ReducerOf<Self> {
    Reduce { state, _ in
      state.value = if true {
        true
      } else {
        false
      }
      return .none
    }
  }
}

```

### Stack dump

```text
0.	Program arguments: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/swift-frontend -frontend -c -primary-file /Users/kaioelfke/Desktop/TCAIfElseExpression/TCAIfElseExpression/ContentView.swift /Users/kaioelfke/Desktop/TCAIfElseExpression/TCAIfElseExpression/TCAIfElseExpressionApp.swift /Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Intermediates.noindex/TCAIfElseExpression.build/Debug-iphoneos/TCAIfElseExpression.build/DerivedSources/GeneratedAssetSymbols.swift -emit-dependencies-path /Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Intermediates.noindex/TCAIfElseExpression.build/Debug-iphoneos/TCAIfElseExpression.build/Objects-normal/arm64/ContentView.d -emit-const-values-path /Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Intermediates.noindex/TCAIfElseExpression.build/Debug-iphoneos/TCAIfElseExpression.build/Objects-normal/arm64/ContentView.swiftconstvalues -emit-reference-dependencies-path /Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Intermediates.noindex/TCAIfElseExpression.build/Debug-iphoneos/TCAIfElseExpression.build/Objects-normal/arm64/ContentView.swiftdeps -serialize-diagnostics-path /Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Intermediates.noindex/TCAIfElseExpression.build/Debug-iphoneos/TCAIfElseExpression.build/Objects-normal/arm64/ContentView.dia -emit-localized-strings -emit-localized-strings-path /Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Intermediates.noindex/TCAIfElseExpression.build/Debug-iphoneos/TCAIfElseExpression.build/Objects-normal/arm64 -target arm64-apple-ios17.2 -Xllvm -aarch64-use-tbi -enable-objc-interop -stack-check -sdk /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS17.2.sdk -I /Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Products/Debug-iphoneos -F /Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Products/Debug-iphoneos/PackageFrameworks -F /Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Products/Debug-iphoneos/PackageFrameworks -F /Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Products/Debug-iphoneos/PackageFrameworks -F /Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Products/Debug-iphoneos/PackageFrameworks -F /Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Products/Debug-iphoneos/PackageFrameworks -F /Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Products/Debug-iphoneos/PackageFrameworks -F /Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Products/Debug-iphoneos/PackageFrameworks -F /Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Products/Debug-iphoneos/PackageFrameworks -F /Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Products/Debug-iphoneos/PackageFrameworks -F /Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Products/Debug-iphoneos/PackageFrameworks -F /Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Products/Debug-iphoneos/PackageFrameworks -F /Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Products/Debug-iphoneos/PackageFrameworks -F /Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Products/Debug-iphoneos -no-color-diagnostics -enable-testing -g -module-cache-path /Users/kaioelfke/Library/Developer/Xcode/DerivedData/ModuleCache.noindex -swift-version 5 -enforce-exclusivity=checked -Onone -D DEBUG -serialize-debugging-options -load-plugin-executable /Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Products/Debug/CasePathsMacros#CasePathsMacros -load-plugin-executable /Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Products/Debug/ComposableArchitectureMacros#ComposableArchitectureMacros -load-plugin-executable /Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Products/Debug/DependenciesMacrosPlugin#DependenciesMacrosPlugin -const-gather-protocols-file /Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Intermediates.noindex/TCAIfElseExpression.build/Debug-iphoneos/TCAIfElseExpression.build/Objects-normal/arm64/TCAIfElseExpression_const_extract_protocols.json -enable-bare-slash-regex -empty-abi-descriptor -validate-clang-modules-once -clang-build-session-file /Users/kaioelfke/Library/Developer/Xcode/DerivedData/ModuleCache.noindex/Session.modulevalidation -Xcc -working-directory -Xcc /Users/kaioelfke/Desktop/TCAIfElseExpression -resource-dir /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift -enable-anonymous-context-mangled-names -Xcc -ivfsstatcache -Xcc /Users/kaioelfke/Library/Developer/Xcode/DerivedData/SDKStatCaches.noindex/iphoneos17.2-21C52-884b7f60ac6761a492c03f282b824eb9.sdkstatcache -Xcc -I/Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Intermediates.noindex/TCAIfElseExpression.build/Debug-iphoneos/TCAIfElseExpression.build/swift-overrides.hmap -Xcc -iquote -Xcc /Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Intermediates.noindex/TCAIfElseExpression.build/Debug-iphoneos/TCAIfElseExpression.build/TCAIfElseExpression-generated-files.hmap -Xcc -I/Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Intermediates.noindex/TCAIfElseExpression.build/Debug-iphoneos/TCAIfElseExpression.build/TCAIfElseExpression-own-target-headers.hmap -Xcc -I/Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Intermediates.noindex/TCAIfElseExpression.build/Debug-iphoneos/TCAIfElseExpression.build/TCAIfElseExpression-all-non-framework-target-headers.hmap -Xcc -ivfsoverlay -Xcc /Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Intermediates.noindex/TCAIfElseExpression.build/Debug-iphoneos/TCAIfElseExpression.build/all-product-headers.yaml -Xcc -iquote -Xcc /Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Intermediates.noindex/TCAIfElseExpression.build/Debug-iphoneos/TCAIfElseExpression.build/TCAIfElseExpression-project-headers.hmap -Xcc -I/Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Products/Debug-iphoneos/include -Xcc -I/Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Intermediates.noindex/TCAIfElseExpression.build/Debug-iphoneos/TCAIfElseExpression.build/DerivedSources-normal/arm64 -Xcc -I/Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Intermediates.noindex/TCAIfElseExpression.build/Debug-iphoneos/TCAIfElseExpression.build/DerivedSources/arm64 -Xcc -I/Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Intermediates.noindex/TCAIfElseExpression.build/Debug-iphoneos/TCAIfElseExpression.build/DerivedSources -Xcc -DDEBUG=1 -module-name TCAIfElseExpression -frontend-parseable-output -disable-clang-spi -target-sdk-version 17.2 -target-sdk-name iphoneos17.2 -external-plugin-path /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS17.2.sdk/usr/lib/swift/host/plugins#/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS17.2.sdk/usr/bin/swift-plugin-server -external-plugin-path /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS17.2.sdk/usr/local/lib/swift/host/plugins#/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS17.2.sdk/usr/bin/swift-plugin-server -external-plugin-path /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/usr/lib/swift/host/plugins#/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/usr/bin/swift-plugin-server -external-plugin-path /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/usr/local/lib/swift/host/plugins#/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/usr/bin/swift-plugin-server -plugin-path /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/host/plugins -plugin-path /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/local/lib/swift/host/plugins -o /Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Build/Intermediates.noindex/TCAIfElseExpression.build/Debug-iphoneos/TCAIfElseExpression.build/Objects-normal/arm64/ContentView.o -index-unit-output-path /TCAIfElseExpression.build/Debug-iphoneos/TCAIfElseExpression.build/Objects-normal/arm64/ContentView.o -index-store-path /Users/kaioelfke/Library/Developer/Xcode/DerivedData/TCAIfElseExpression-bmqctoidpaeivvearwtlubuaahfz/Index.noindex/DataStore -index-system-modules
1.	Apple Swift version 5.9.2 (swiftlang-5.9.2.2.56 clang-1500.1.0.2.5)
2.	Compiling with the current language version
3.	While evaluating request TypeCheckSourceFileRequest(source_file "/Users/kaioelfke/Desktop/TCAIfElseExpression/TCAIfElseExpression/ContentView.swift")
4.	While evaluating request TypeCheckFunctionBodyRequest(TCAIfElseExpression.(file).Feature._@/Users/kaioelfke/Desktop/TCAIfElseExpression/TCAIfElseExpression/ContentView.swift:11:34)
5.	While type-checking statement at [/Users/kaioelfke/Desktop/TCAIfElseExpression/TCAIfElseExpression/ContentView.swift:12:12 - line:19:5] RangeText="{ state, _ in
      state.value = if true {
        true
      } else {
        false
      }
      return .none
    "
6.	While type-checking expression at [/Users/kaioelfke/Desktop/TCAIfElseExpression/TCAIfElseExpression/ContentView.swift:13:7 - line:17:7] RangeText="state.value = if true {
        true
      } else {
        false
      "
7.	While type-checking-target starting at /Users/kaioelfke/Desktop/TCAIfElseExpression/TCAIfElseExpression/ContentView.swift:13:19
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  swift-frontend           0x0000000106d55abc llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 56
1  swift-frontend           0x00000001099bbcb0 llvm::sys::RunSignalHandlers() + 112
2  swift-frontend           0x0000000109725054 SignalHandler(int) + 352
3  libsystem_platform.dylib 0x0000000188681a24 _sigtramp + 56
4  swift-frontend           0x0000000106354030 (anonymous namespace)::SyntacticElementSolutionApplication::SyntacticElementSolutionApplication(swift::constraints::Solution&, (anonymous namespace)::SyntacticElementContext, std::__1::function<llvm::Optional<swift::constraints::SyntacticElementTarget> (swift::constraints::SyntacticElementTarget)>) + 556
5  swift-frontend           0x0000000106354030 (anonymous namespace)::SyntacticElementSolutionApplication::SyntacticElementSolutionApplication(swift::constraints::Solution&, (anonymous namespace)::SyntacticElementContext, std::__1::function<llvm::Optional<swift::constraints::SyntacticElementTarget> (swift::constraints::SyntacticElementTarget)>) + 556
6  swift-frontend           0x00000001084e2da0 swift::constraints::ConstraintSystem::applySolutionToSingleValueStmt(swift::constraints::Solution&, swift::SingleValueStmtExpr*, swift::DeclContext*, std::__1::function<llvm::Optional<swift::constraints::SyntacticElementTarget> (swift::constraints::SyntacticElementTarget)>) + 144
7  swift-frontend           0x00000001083f22a0 (anonymous namespace)::ExprWalker::walkToExprPre(swift::Expr*) + 3588
8  swift-frontend           0x0000000105a1ca2c (anonymous namespace)::Traversal::visitAssignExpr(swift::AssignExpr*) + 168
9  swift-frontend           0x00000001045d0020 (anonymous namespace)::Traversal::visit(swift::Expr*) (.llvm.16965159650233048978) + 1052
10 swift-frontend           0x00000001083e288c (anonymous namespace)::ExprWalker::rewriteTarget(swift::constraints::SyntacticElementTarget) + 2688
11 swift-frontend           0x00000001062dd79c swift::constraints::ConstraintSystem::applySolution(swift::constraints::Solution&, swift::constraints::SyntacticElementTarget) + 6956
12 swift-frontend           0x0000000106921b1c swift::TypeChecker::typeCheckTarget(swift::constraints::SyntacticElementTarget&, swift::OptionSet<swift::TypeCheckExprFlags, unsigned int>) + 756
13 swift-frontend           0x00000001090bae4c swift::TypeChecker::typeCheckExpression(swift::constraints::SyntacticElementTarget&, swift::OptionSet<swift::TypeCheckExprFlags, unsigned int>) + 436
14 swift-frontend           0x00000001096e1460 (anonymous namespace)::StmtChecker::typeCheckASTNode(swift::ASTNode&) (.llvm.3900483966656927864) + 396
15 swift-frontend           0x0000000106bd8840 swift::ASTVisitor<(anonymous namespace)::StmtChecker, void, swift::Stmt*, void, void, void, void>::visit(swift::Stmt*) (.llvm.3900483966656927864) + 296
16 swift-frontend           0x00000001096ec9d8 bool (anonymous namespace)::StmtChecker::typeCheckStmt<swift::BraceStmt>(swift::BraceStmt*&) (.llvm.3900483966656927864) + 316
17 swift-frontend           0x0000000106bd1bbc swift::TypeChecker::typeCheckClosureBody(swift::ClosureExpr*) + 324
18 swift-frontend           0x00000001062dd8c0 swift::constraints::ConstraintSystem::applySolution(swift::constraints::Solution&, swift::constraints::SyntacticElementTarget) + 7248
19 swift-frontend           0x00000001062c28cc swift::TypeChecker::applyResultBuilderBodyTransform(swift::FuncDecl*, swift::Type, bool) + 2672
20 swift-frontend           0x00000001096e8f10 swift::TypeCheckFunctionBodyRequest::evaluate(swift::Evaluator&, swift::AbstractFunctionDecl*) const + 404
21 swift-frontend           0x00000001072d4e58 swift::TypeCheckFunctionBodyRequest::OutputType swift::evaluateOrDefault<swift::TypeCheckFunctionBodyRequest>(swift::Evaluator&, swift::TypeCheckFunctionBodyRequest, swift::TypeCheckFunctionBodyRequest::OutputType) + 884
22 swift-frontend           0x0000000106c829a8 swift::TypeCheckSourceFileRequest::evaluate(swift::Evaluator&, swift::SourceFile*) const + 608
23 swift-frontend           0x0000000106c88658 llvm::Expected<swift::TypeCheckSourceFileRequest::OutputType> swift::Evaluator::getResultUncached<swift::TypeCheckSourceFileRequest>(swift::TypeCheckSourceFileRequest const&) + 664
24 swift-frontend           0x0000000106c66a4c swift::TypeCheckSourceFileRequest::OutputType swift::evaluateOrDefault<swift::TypeCheckSourceFileRequest>(swift::Evaluator&, swift::TypeCheckSourceFileRequest, swift::TypeCheckSourceFileRequest::OutputType) + 228
25 swift-frontend           0x000000010733da8c swift::CompilerInstance::performSema() + 160
26 swift-frontend           0x0000000109298504 performCompile(swift::CompilerInstance&, int&, swift::FrontendObserver*) + 464
27 swift-frontend           0x000000010929c854 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 4568
28 swift-frontend           0x0000000109303d44 swift::mainEntry(int, char const**) + 4408
29 dyld                     0x00000001882d10e0 start + 2360
```

### Expected behavior

It should compile.

### Environment

Xcode Version 15.2 (15C500b)
macOS 14.2.1 (23C71)

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by calda**
This PR fixes an issue with the implementation of SE-0365, where the following code would unexpectedly emit an error:

```swift
@resultBuilder
struct VoidBuilder {
    static func buildBlock() -> Void { }
    static func buildPartialBlock<T>(first: T) -> Void { }
    static func buildPartialBlock<T>(accumulated: Void, next: T) -> Void { }
}

final class EscapingWrapper {
    static func wrapper(_ closure: @escaping () -> Void) {
        closure()
    }
}

final class Sample {
    @VoidBuilder
    var void: Void {
        EscapingWrapper.wrapper { [weak self] in
            guard let self else { return } // Unexpected error: Implicit use of 'self' in closure; use 'self.' to make capture semantics explicit
        }
    }
}
```

Fixes #64757, fixes #71040.

Please review: @xedin

### CODE REVIEW

File: lib/Sema/MiscDiagnostics.cpp [hunk 1]
```diff
@@ -1991,6 +2025,18 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
...
   }
+  
+  // Verify E is actually a descendant of ACE before we use it.
+  // Otherwise we'd be diagnosing E within the context of a closure
+  // that it isn't actually a part of.
+  if (ACE) {
+    auto isDescendantWalker = ExprIsDescendantWalker(const_cast<Expr *>(E));
+    ACE->walk(isDescendantWalker);
+    if (!isDescendantWalker.exprIsDescendant) {
+      ACE = nullptr;
```
Comment (calda) [inline lib/Sema/MiscDiagnostics.cpp:2036]: The implementation of `diagnoseImplicitSelfUseInClosure` effectively requires `E` to be a part of the body of `ACE`. 

For some reason, in this specific case using result builders, this code was being invoked with an expression that wasn't actually a descendant of the closure.

A fix that seems to work well here is to just check whether or not `E` is a descendant of `ACE` by walking the AST of the closure and checking if we encounter `E`.

**Comment by xedin** [comment]
@ahoppen Looks like this is the issue here https://github.com/apple/swift/blob/main/lib/Sema/ConstraintSystem.cpp#L7112. Should this be set only in code completion mode?

**Comment by calda** [comment]
Thanks for the review and suggestion @xedin. 

In https://github.com/apple/swift/pull/64786/commits/10878e8aba614e0b10c837e8bdd2efc83e87a74a I updated the condition you sent to also include `isForCodeCompletion()` -- is that about what you had in mind?

The tests seem to pass locally after I made this change.

**Comment by xedin** [comment]
@calda I haven't look at where this flag gets set but I was hoping that this check you added code be moved there so we end up without this option if solver is not in code completion mode...

**Comment by calda** [comment]
Thanks @xedin, that makes sense. I moved the change up to the callsite of the method that configures that option. All the tests pass still with this change.

**Review by xedin — APPROVED** [review]

**Comment by xedin** [comment]
@ahoppen WDYT?

**Comment by xedin** [comment]
@swift-ci please test

File: lib/Sema/TypeCheckStmt.cpp [hunk 2]
```diff
@@ -2649,7 +2649,9 @@ TypeCheckFunctionBodyRequest::evaluate(Evaluator &evaluator,
     if (Type builderType = getResultBuilderType(func)) {
       if (auto optBody =
               TypeChecker::applyResultBuilderBodyTransform(
-                func, builderType)) {
+                func, builderType,
+                /*ClosuresInResultBuilderDontParticipateInInference=*/
+                false)) {
```
Comment (xedin) [inline lib/Sema/TypeCheckStmt.cpp:2654]: @ahoppen Should we flip default to `false` maybe instead?

**Comment by calda** [comment]
Sure @xedin, here you go: https://github.com/apple/swift/pull/64786/commits/7c78ad2860dc96d53d1386e3eb83f497c7bc8b2b

**Comment by xedin** [comment]
@swift-ci Please SourceKit stress test

**Comment by xedin** [comment]
@swift-ci please test

**Pull request merged**